File: src/lib/services/system/updates.ts
import { invoke } from '@tauri-apps/api/core';
import { log } from '@/utils/logger';
import { get } from 'svelte/store';
import { settings } from '@/features/settings';

export interface AppVersionInfo {
  currentVersion: string;
  checkedAt: number;
}

const GITHUB_RELEASES_BASE = 'https://github.com/bismuth-pkm/bismuth/releases/download';

const UPDATE_ENDPOINTS: Record<string, string> = {
  alpha: `${GITHUB_RELEASES_BASE}/alpha-latest/latest.json`,
  beta: `${GITHUB_RELEASES_BASE}/beta-latest/latest.json`,
  release: `${GITHUB_RELEASES_BASE}/release-latest/latest.json`,
};

/**
 * Returns the updater endpoint URL for the given channel.
 * Currently stub — real server not yet deployed; returns the channel URL string.
 * Unknown values fall back to the release endpoint.
 */
export function buildUpdateEndpoint(channel: 'alpha' | 'beta' | 'release' | string): string {
  return UPDATE_ENDPOINTS[channel] ?? UPDATE_ENDPOINTS['release'];
}

/**
 * Calls the Rust set_app_locale stub to log the locale code.
 * Part of the i18n call chain required by T065. Actual translation wiring is future scope.
 */
export async function setAppLocale(locale: string): Promise<void> {
  try {
    await invoke('set_app_locale', { locale });
    log.debug('App locale set', { locale });
  } catch (error) {
    log.warn('set_app_locale failed (non-blocking)', { locale, error: String(error) });
  }
}

/**
 * Returns the current app version from the Tauri package info.
 * Does not make any network calls.
 */
export async function checkAppVersion(): Promise<AppVersionInfo & { channel: string }> {
  const channel = get(settings).updateChannel ?? 'release';
  const endpoint = buildUpdateEndpoint(channel);
  log.debug('checkAppVersion', { channel, endpoint });
  try {
    const result = await invoke<{ current_version: string; checked_at: number }>('check_app_version');
    return { currentVersion: result.current_version, checkedAt: result.checked_at, channel };
  } catch (error) {
    log.error('Failed to check app version', error as Error);
    throw error;
  }
}

File: src-tauri/src/commands/system/updates.rs
//! App version and update-check stub commands.
//! No network calls are made — this command reads only the compiled-in app version.
//! A real update endpoint may be wired in future; at that point, responses MUST
//! be verified against a signed release manifest before display.

use serde::{Deserialize, Serialize};
use tauri::AppHandle;

#[derive(Debug, Serialize, Deserialize)]
pub struct AppVersionInfo {
    pub current_version: String,
    pub checked_at: i64,
}

/// Sets the app locale. Stub — logs the locale code for the future i18n call chain.
/// Validates that the locale is non-empty and at most 16 chars.
#[tauri::command]
pub async fn set_app_locale(locale: String) -> Result<(), String> {
    if locale.is_empty() || locale.len() > 16 {
        return Err("Invalid locale code".into());
    }
    tracing::info!("set_app_locale: locale={}", locale);
    Ok(())
}

/// Returns the current app version from the Tauri package info.
/// Does not make any network calls.
#[tauri::command]
pub async fn check_app_version(app: AppHandle) -> Result<AppVersionInfo, String> {
    let version = app.package_info().version.to_string();
    let checked_at = chrono::Utc::now().timestamp();
    tracing::info!("check_app_version: current={}", version);
    Ok(AppVersionInfo { current_version: version, checked_at })
}
